### ISSUE TITLE
Should astro allow for Node.js builtin access? (import 'fs')

### ISSUE THREAD

**Issue by FredKSchott** [snowpackjs/astro#341]
## Background

- Currently, we prevent you from accessing Node.js builtins in Astro
- This is partially because that's Snowpack's default behavior, but also because we wanted to keep Astro agnostic to any single runtime if we could help it (ex: to target a platform like Cloudflare Workers in the future).

## Feature Request

**Use-case:** In Snowpack.dev, I want to read the package.json version from disk and then show that in the site header (as.seen on snowpack.dev). Currently, this is impossible (errors if you try to `import 'fs'`).

```js
import fs from 'fs';

const snowpackManifest = JSON.parse(
  fs.readFileSync(path.join(__dirname, '../../snowpack/package.json'), 'utf8'),
);

export default snowpackManifest.version;
```

## Proposed Solutions

1. Allow `import 'fs'` and others. If you use them, then you take on the responsibility of being Node-specific. 
2. Add some special `src/data/` file type that will always run server-side during build and has access to Node.js builtins. Not possible to import these in a route.
3. Disallow generally, and force you to move this kind of code into an `astro.config.mjs` file where it can be injected into the build as a static value. 

## Prior art: 

- (Solution 1) I can't find a fullstack framework that lets you do `import 'fs'` in a route.
- (Solution 2) Eleventy recently added "Data JS Files": https://www.11ty.dev/docs/data-js/ This is a special file type that must live in a special place and Eleventy will run them at build time.
- (Solution 3) Before that, you would need to load this data inside of your config file and inject it into your build that way.

**Comment by brandonpittman** [comment]
You can do this in Next.js. This case is handled by getStaticProps. 

Am I misreading the problem?

**Comment by trustedtomato** [comment]
In my case, I would like to access `fs` and `gm` (which uses other Node.js built-ins) to process the static images and create a downscaled blurred placeholder for them. This would be pretty easy if one had access to Node.js globals and built-ins.
The front-matter-scripts in the routes can only run server-side, so I think it would make a lot of sense if they had access to the runtime's globals and modules.

Connected to prior art:
1. SvelteKit surely allows accessing arbitrary Node.js modules too on the server-side.

Maybe there could be a directory which shouldn't be Snowpacked (like `src/no-transform/`) and a route file could import from that directory. Plus, a prefix (like `no-transform:`) could be introduced to import statements to tell Snowpack not to pack those packages (ex. `import foo from 'no-transform:../no-transform/foo.js'` or `import('no-transform:fs')`).

**Comment by FredKSchott** [comment]
We have a limitation here: Snowpack won't let you import 1 thing in SSR mode and another in the browser. Until that's fixed, we need a solution that works for both.

I'm tempted to say lets just enable this for everything. If you try to import a node builtin on the client, it will obviously error (since we wouldn't be touching them anymore) and then at least it would be up to you to fix. Then, we could handle this better once the Snowpack engine is improved.

I think it's too drastic of a missing feature to leave out completely until then.

@matthewp thoughts?

**Comment by matthewp** [comment]
Yeah that's fine. I'd be a little worried about people pulling in large polyfill in the browser if we did it for them. If we're going to advertise minimal JS we need to avoid footguns. But if we're not helping them do that then I think it's ok. Some build warnings on large page sizes could help do (different feature)

**Comment by FredKSchott** [comment]
SGTM /cc @snowpackjs/maintainers for any thoughts, otherwise lets consider this the plan

**Comment by FredKSchott** [comment]
To experiment with: `import ... from 'node:XXX';

**Comment by matthewp** [comment]
I have a fix for this but need two changes in snowpack:

1. Snowpack SSR provides an `import.meta.url` as `/_astro/src/pages/index.astro`. This should be the filepath the same way as if you use ESM in Node directly.
2. Snowpack isn't able to dynamically load `node:` modules at this time.

**Comment by matthewp** [comment]
Depends on: https://github.com/snowpackjs/snowpack/pull/3472

**Comment by matthewp** [comment]
Here's a branch with this tested once snowpack change is in: https://github.com/snowpackjs/astro/compare/allow-node?expand=1

**Comment by natemoo-re** [comment]
@matthewp Realized one case that the explicit `node:` prefix doesn't cover is... dependencies that rely on modules like `fs`. Should Astro setup aliases that prefix all node builtins to the `node:` namespace? That would handle sub-package imports and keep the `node:` requirement for users.

**Comment by matthewp** [comment]
@natemoo-re I hadn't considered that, I'll give it a try!

**Comment by natemoo-re** [comment]
Decided this is the right direction, Snowpack should allow the `node:` prefix _and_ the bare module specifier, Astro compiler will warn if `node:` prefix is missing.

**Comment by FredKSchott** [comment]
Note: `node:` prefix added in https://nodejs.org/en/blog/release/v14.13.1/ & https://nodejs.org/en/blog/release/v12.20.0/

**Comment by matthewp** [comment]
There's a bug in snowpack that prevents packages from using `node:` prefix. Going to fix that before submitting a PR for this.

**Comment by matthewp** [comment]
Released in Astro 0.14.1
